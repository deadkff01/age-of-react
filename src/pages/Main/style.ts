import styled from "styled-components";

export const CivilizationsList = styled.div`
  width: 800px;
  margin: 0 auto;
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  text-align: center;
`;
