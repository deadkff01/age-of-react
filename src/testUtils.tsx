import React from "react";
import { render } from "@testing-library/react";
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router";

export function renderWithRouterWrapper(
  ui: any,
  { route = "/", history = createHistory(createMemorySource(route)) }: any = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    history,
  };
}
