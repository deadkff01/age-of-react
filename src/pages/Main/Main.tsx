import React, { FC, useState, useEffect, useCallback } from "react";
import { getCivilizations } from "services/ageService";
import { tw } from "twind";
import { CivilizationsList } from "./style";

const Main: FC = () => {
  const [civilizations, setCivilizations] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);

  const requestCivilizations = useCallback(async () => {
    try {
      setIsLoading(true);
      const { data } = await getCivilizations();
      //console.log(data);
      setCivilizations(data.civilizations);
    } catch (e) {
      setError(true);
    } finally {
      setIsLoading(false);
    }
  }, []);

  useEffect(() => {
    requestCivilizations();
  }, [requestCivilizations]);

  return (
    <div id="container-main" data-testid="container-main">
      <h1 className={tw`text-center text-5xl font-bold my-10`}>Age of React</h1>
      {civilizations.length > 0 ? (
        <CivilizationsList
          id="civilizations-list"
          data-testid="civilizations-list"
        >
          {civilizations.map((c: any) => (
            <div key={c.id}>{c.name}</div>
          ))}
        </CivilizationsList>
      ) : null}
      {error ? (
        <div className={tw`text-center`} data-testid="error">
          Error
        </div>
      ) : null}
      {isLoading ? (
        <div className={tw`text-center`} data-testid="loading">
          Loading...
        </div>
      ) : null}
    </div>
  );
};

export default Main;
