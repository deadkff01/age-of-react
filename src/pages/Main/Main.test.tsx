import React from "react";
import { screen } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { renderWithRouterWrapper } from "../../testUtils";
import Main from "./Main";
import { civilizationsMock } from "mocks/civilizations";
import { act } from "react-dom/test-utils";

const server = setupServer(
  rest.get(`${process.env.REACT_APP_BACKEND}/civilizations`, (_, res, ctx) => {
    return res(ctx.json(civilizationsMock));
  })
);

describe("Main page", () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  test("render the page", async () => {
    renderWithRouterWrapper(<Main />);
    const container = await screen.findByTestId("civilizations-list");
    expect(container).toMatchSnapshot();
  });

  test("civilization length", async () => {
    renderWithRouterWrapper(<Main />);
    const civilizationsList = await screen.findByTestId("civilizations-list");
    expect(civilizationsList.children.length).toBe(32);
  });

  test("error", async () => {
    server.use(
      rest.get(
        `${process.env.REACT_APP_BACKEND}/civilizations`,
        (_, res, ctx) => {
          return res(ctx.status(500));
        }
      )
    );

    renderWithRouterWrapper(<Main />);
    const error = await screen.findByTestId("error");
    expect(error).toBeDefined();
  });

  test("loading", async () => {
    renderWithRouterWrapper(<Main />);
    await act(async () => {
      jest.useFakeTimers();
    });
    const loading = await screen.findByTestId("loading");
    expect(loading).toBeDefined();
  });
});
