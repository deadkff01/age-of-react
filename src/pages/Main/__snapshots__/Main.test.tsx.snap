// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page render the page 1`] = `
<div
  class="sc-bdnxRM ikrJdP"
  data-testid="civilizations-list"
>
  <div>
    Aztecs
  </div>
  <div>
    Britons
  </div>
  <div>
    Byzantines
  </div>
  <div>
    Celts
  </div>
  <div>
    Chinese
  </div>
  <div>
    Franks
  </div>
  <div>
    Goths
  </div>
  <div>
    Huns
  </div>
  <div>
    Japanese
  </div>
  <div>
    Koreans
  </div>
  <div>
    Mayans
  </div>
  <div>
    Mongols
  </div>
  <div>
    Persians
  </div>
  <div>
    Saracens
  </div>
  <div>
    Spanish
  </div>
  <div>
    Teutons
  </div>
  <div>
    Turks
  </div>
  <div>
    Vikings
  </div>
  <div>
    Berbers
  </div>
  <div>
    Burmese
  </div>
  <div>
    Ethiopians
  </div>
  <div>
    Goths
  </div>
  <div>
    Incas
  </div>
  <div>
    Indians
  </div>
  <div>
    Italians
  </div>
  <div>
    Japanese
  </div>
  <div>
    Khmer
  </div>
  <div>
    Magyars
  </div>
  <div>
    Malians
  </div>
  <div>
    Portuguese
  </div>
  <div>
    Slavs
  </div>
  <div>
    Vietnamese
  </div>
</div>
`;
